src/rust/include/rust/orchard/keys.h

```diff
@@ -0,0 +1,216 @@
+ */
+bool orchard_incoming_viewing_key_eq(
+    const OrchardIncomingViewingKeyPtr* k0,
+    const OrchardIncomingViewingKeyPtr* k1);
+
+//
+// Full Viewing Keys
+//
+
+struct OrchardFullViewingKeyPtr;
+typedef struct OrchardFullViewingKeyPtr OrchardFullViewingKeyPtr;
+
+/**
+ * Clones the given Orchard full viewing key and returns
+ * a pointer to the newly created value. Both the original
+ * one's memory and the newly allocated one need to be freed
+ * independently.
+ */
+OrchardFullViewingKeyPtr* orchard_full_viewing_key_clone(
+        const OrchardFullViewingKeyPtr* ptr);
+
+/**
+ * Free the memory allocated for the given Orchard full viewing key
+ */
+void orchard_full_viewing_key_free(OrchardFullViewingKeyPtr* ptr);
+
+/**
+ * Parses an Orchard full viewing key from the given stream.
+ *
+ * - If the key does not parse correctly, the returned pointer will be null.
+ */
+OrchardFullViewingKeyPtr* orchard_full_viewing_key_parse(
+    void* stream,
+    read_callback_t read_cb);
+
+/**
+ * Serializes an Orchard full viewing key to the given stream
+ *
+ * If `full_viewing_key == nullptr`, this serializes `nActionsOrchard = 0`.
+ */
+bool orchard_full_viewing_key_serialize(
+    const OrchardFullViewingKeyPtr* full_viewing_key,
+    void* stream,
+    write_callback_t write_cb);
+
+/**
+ * Returns the incoming viewing key for the specified full viewing key.
+ */
+OrchardIncomingViewingKeyPtr* orchard_full_viewing_key_to_incoming_viewing_key(
+    const OrchardFullViewingKeyPtr* key);
+
+/**
+ * Implements equality testing between full viewing keys.
+ */
+bool orchard_full_viewing_key_eq(
+    const OrchardFullViewingKeyPtr* k0,
+    const OrchardFullViewingKeyPtr* k1);
+
+//
+// Spending Keys
+//
+
+struct OrchardSpendingKeyPtr;
+typedef struct OrchardSpendingKeyPtr OrchardSpendingKeyPtr;
+
+/**
+ * Constructs a spending key by ZIP-32 derivation to the account
+ * level from the provided seed.
+ */
+OrchardSpendingKeyPtr* orchard_spending_key_for_account(
+        const unsigned char* seed,
+        size_t seed_len,
+        uint32_t bip44_coin_type,
+        uint32_t account_id);
+
+/**
+ * Clones the given Orchard spending key and returns
+ * a pointer to the newly created value. Both the original
+ * one's memory and the newly allocated one need to be freed
+ * independently.
+ */
+OrchardSpendingKeyPtr* orchard_spending_key_clone(
+        const OrchardSpendingKeyPtr* ptr);
+
+/**
+ * Free the memory allocated for the given Orchard spending key
+ */
+void orchard_spending_key_free(OrchardSpendingKeyPtr* ptr);
+
+/**
+ * Parses an Orchard spending key from the given stream.
+ *
+ * - If the key does not parse correctly, the returned pointer will be null.
+ */
+OrchardSpendingKeyPtr* orchard_spending_key_parse(
+    void* stream,
+    read_callback_t read_cb);
+
+/**
+ * Serializes an Orchard spending key to the given stream
+ *
+ * If `spending_key == nullptr`, this will return `false`
```

Comment:
This method actually panics if `spending_key == nullptr`, so that must not occur (as with many other FFI methods).
```suggestion
 * Serializes an Orchard spending key to the given stream.
```